core/runtime/impl/block_builder_impl.cpp

```diff
@@ -52,73 +47,68 @@ namespace kagome::runtime {
 
     runtime::SizeType data_size = enc_data.size();
     // TODO (Harrm) PRE-98: after check for memory overflow is done, refactor it
     runtime::WasmPointer ptr = memory_->allocate(data_size);
-    memory_->storeBuffer(ptr, enc_data);
+    memory_->storeBuffer(ptr, common::Buffer(std::move(enc_data)));
 
     wasm::LiteralList ll{Literal(ptr), Literal(data_size)};
 
     OUTCOME_TRY(
         res,
         executor_.call(state_code_, "BlockBuilder_inherent_extrinsics", ll));
 
-    uint32_t res_addr = getWasmAddr(res.geti64());
+    WasmPointer res_addr = getWasmAddr(res.geti64());
+    SizeType len = getWasmLen(res.geti64());
 
-    WasmMemoryStream s(memory_);
-    OUTCOME_TRY(s.advance(res_addr));
+    auto buffer = memory_->loadN(res_addr, len);
+    ScaleDecoderStream s(buffer);
 
-    auto decode_f = [this](common::ByteStream &stream) {
-      return codec_->decodeExtrinsic(stream);
-    };
-    OUTCOME_TRY(decoded_res, decodeCollection<Extrinsic>(s, decode_f));
+    OUTCOME_TRY(decoded_res, scale::decode<std::vector<Extrinsic>>(s));
 
     return std::move(decoded_res);
   }
 
   outcome::result<CheckInherentsResult> BlockBuilderImpl::check_inherents(
       const primitives::Block &block, const primitives::InherentData &data) {
-    OUTCOME_TRY(enc_block, codec_->encodeBlock(block));
-    OUTCOME_TRY(enc_data, codec_->encodeInherentData(data));
+    ScaleEncoderStream os;
+    OUTCOME_CATCH((os << block << data))
+    auto encoded_data = Buffer(os.data());
 
     // TODO (Harrm) PRE-98: after check for memory overflow is done, refactor it
-    runtime::SizeType block_size = enc_block.size();
-    runtime::WasmPointer block_ptr = memory_->allocate(block_size);
-    memory_->storeBuffer(block_ptr, enc_block);
+    runtime::SizeType param_size = encoded_data.size();
+    runtime::WasmPointer param_ptr = memory_->allocate(param_size);
+    memory_->storeBuffer(param_ptr, encoded_data);
 
-    runtime::SizeType data_size = enc_data.size();
-    runtime::WasmPointer data_ptr = memory_->allocate(data_size);
-    memory_->storeBuffer(data_ptr, enc_data);
-
-    // TODO(Harrm) PRE-154 check what exactly its input arguments should be
-    wasm::LiteralList ll{Literal(data_ptr), Literal(data_size)};
+    wasm::LiteralList ll{Literal(param_ptr), Literal(param_size)};
 
     OUTCOME_TRY(
         res, executor_.call(state_code_, "BlockBuilder_check_inherents", ll));
 
-    uint32_t res_addr = getWasmAddr(res.geti64());
+    WasmPointer res_addr = getWasmAddr(res.geti64());
+    SizeType len = getWasmLen(res.geti64());
 
-    WasmMemoryStream s(memory_);
-    OUTCOME_TRY(s.advance(res_addr));
+    auto buffer = memory_->loadN(res_addr, len);
+    ScaleDecoderStream s(buffer);
 
-    scale::TypeDecoder<bool> decoder;
-    OUTCOME_TRY(ok, decoder.decode(s));
-    OUTCOME_TRY(is_fatal, decoder.decode(s));
-    OUTCOME_TRY(error_data, codec_->decodeInherentData(s));
+    OUTCOME_TRY(ok, scale::decode<bool>(s));
```

Comment:
rework with just scale::decode< CheckInherentsResult >(buffer);

Reply:
added operator>> for CheckInherentsResult and reworked decoding in api